src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs

```diff
@@ -4380,27 +4380,38 @@ internal static IEnumerable<CompletionResult> CompleteFilename(CompletionContext
             var quote = HandleDoubleAndSingleQuote(ref wordToComplete);
             var results = new List<CompletionResult>();
 
-            // First, try to match \\server\share
-            // support both / and \ when entering UNC paths for typing convenience (#17111)
-            var shareMatch = Regex.Match(wordToComplete, @"^(?:\\\\|//)([^\\/]+)(?:\\|/)([^\\/]*)$");
+            // Matches file shares with and without the provider name and with either slash direction.
+            // Avoids matching Windows device paths like \\.\CDROM0 and \\?\Volume{b8f3fc1c-5cd6-4553-91e2-d6814c4cd375}\
+            var shareMatch = Regex.Match(
```

Comment:
PathIsUnc will return false if the filesystem provider is specified like this: `Microsoft.PowerShell.Core.Filesystem::\\localhost\c$`. So we can't use it as is.  
Also, It returns true when pointing to a file which doesn't match the current design of only matching up to the share portion.

Reply:
I don't know how often people use it but when they do we shouldn't be removing it from the completion result. You can specify the filesystem provider with or without the module name so if we want to avoid Regex then we need 2 "StartsWith" tests on the string along with the UNC check and the check to see if it's a file or a fileshare that the UNC path is pointing to.  
Maybe I'm underestimating the impact of regex but I don't feel like replacing 1 regex pattern with 4 other checks is worth doing.

Reply:
Since this completion may be used often, we should compile and store a static version of the regex.